### PR TITLE
[TIMOB-19708, TIMOB-19709, TIMOB-19710, TIMOB-19711, TIMOB-19712, TIMOB-19713, TIMOB-19714, TIMOB-19715] Fix and improve UIApplicationShortcuts

### PR DESCRIPTION
**Pull request by hansemannn**
JIRA tickets:
- https://jira.appcelerator.org/browse/TIMOB-19708
- https://jira.appcelerator.org/browse/TIMOB-19709
- https://jira.appcelerator.org/browse/TIMOB-19710
- https://jira.appcelerator.org/browse/TIMOB-19711
- https://jira.appcelerator.org/browse/TIMOB-19712
- https://jira.appcelerator.org/browse/TIMOB-19713
- https://jira.appcelerator.org/browse/TIMOB-19714
- https://jira.appcelerator.org/browse/TIMOB-19715

Updated demo: https://gist.github.com/hansemannn/c6daf98c25fd7635d1cf

/cc @FokkeZB @benbahrenburg

### CODE REVIEW

File: iphone/iphone/Titanium.plist [hunk 1]
```diff
@@ -57,5 +57,23 @@
 		<string>UIInterfaceOrientationLandscapeLeft</string>
 		<string>UIInterfaceOrientationLandscapeRight</string>
 	</array>
+    <key>UIApplicationShortcutItems</key>
```
Comment (FokkeZB) [inline iphone/iphone/Titanium.plist]: Uh.. you don't actually want this in the plist template right? ;)

File: iphone/Classes/TiUIiOSApplicationShortcutsProxy.m [hunk 2]
```diff
@@ -173,19 +204,44 @@ -(void)addShortcutItem:(id)args
...
-        return [UIApplicationShortcutIcon iconWithTemplateImageName:(NSString*)value];
+        return [UIApplicationShortcutIcon iconWithTemplateImageName:[self urlInAssetCatalog:value]];
     }
     
-    NSLog(@"[ERROR] Invalid icon selection provided, defaulting to SHORTCUT_ICON_TYPE_COMPOSE");
+    NSLog(@"[ERROR] Invalid icon provided, defaulting to SHORTCUT_ICON_TYPE_COMPOSE.");
     return UIApplicationShortcutIconTypeCompose;
 }
 
+-(NSString*)urlInAssetCatalog:(NSString*)url
```
Comment (FokkeZB) [inline iphone/Classes/TiUIiOSApplicationShortcutsProxy.m]: Similar stuff happens here:
https://github.com/appcelerator/titanium_mobile/pull/7248/files

Maybe should be in one util class?

File: apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml [hunk 3]
```diff
@@ -30,112 +31,169 @@ methods:
...
 
-examples: 
+examples:
   - title: Example app.js
     example: |
         The following code excerpt demonstrates
 
-            var win = Titanium.UI.createWindow({  
-                title:'Example', backgroundColor:'#fff', layout:"vertical"
+            Ti.App.addEventListener("resume", function(){
```
Comment (FokkeZB) [inline apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml]: Why the resume and resumed here?

File: apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml [hunk 4]
```diff
@@ -30,112 +31,169 @@ methods:
...
+
+            win.open();
+
+---
+name: ShortcutParams
+summary: Dictionary of options for <Titanium.UI.iOS.addDynamicShortcut>.
+description: |
+    The parameters used when creating the dynamic shortcut.  Must include the `itemtype` and `title` properties.
+properties:
+  - name: type
```
Comment (FokkeZB) [inline apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml]: should be `itemtype`

File: apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml [hunk 5]
```diff
@@ -30,112 +31,169 @@ methods:
...
-            	* icon - the icon to be displayed on the application shortcut, for example Titanium.UI.iOS.SHORTCUT_ICON_TYPE_COMPOSE
-        type: Dictionary
-
-  - name: removeShortcutItem
-    summary: Removes the provided application shortcut item.
+            Use the `itemtype` property to determine which shortcut should be removed.
+        type: String
+
+  - name: getDynamicShortcut
+    summary: Gets the dynamic application shortcut item identified by the `itemtype`.
```
Comment (FokkeZB) [inline apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml]: Should add that it is "read-only" - that changing the returned object will not have any effect.

File: iphone/Classes/TiUIiOSApplicationShortcutsProxy.m [hunk 6]
```diff
@@ -173,19 +204,44 @@ -(void)addShortcutItem:(id)args
...
+    NSData *stringBytes = [url dataUsingEncoding: NSUTF8StringEncoding];
+    if (CC_SHA1([stringBytes bytes], (CC_LONG)[stringBytes length], digest)) {
+        // SHA-1 hash has been calculated and stored in 'digest'.
+        NSMutableString *sha = [[NSMutableString alloc] init];
+        for (int i = 0; i < CC_SHA1_DIGEST_LENGTH; i++) {
+            [sha appendFormat:@"%02x", digest[i]];
+        }
+        [sha appendString:@"."];
+        [sha appendString:[url pathExtension]];
+        resultUrl = sha;
```
Comment (FokkeZB) [inline iphone/Classes/TiUIiOSApplicationShortcutsProxy.m]: This should be:

```
resultUrl = [NSMutableString stringWithString: sha];
```

else nilling `sha` on the next line will nill `resultUrl` as well.

File: iphone/Classes/TiUIiOSApplicationShortcutsProxy.m [hunk 7]
```diff
@@ -173,19 +204,44 @@ -(void)addShortcutItem:(id)args
...
+    NSLog(@"[ERROR] Invalid icon provided, defaulting to SHORTCUT_ICON_TYPE_COMPOSE.");
     return UIApplicationShortcutIconTypeCompose;
 }
 
+-(NSString*)urlInAssetCatalog:(NSString*)url
+{
+    NSString *resultUrl = nil;
+    
+    unsigned char digest[CC_SHA1_DIGEST_LENGTH];
+    NSData *stringBytes = [url dataUsingEncoding: NSUTF8StringEncoding];
```
Comment (FokkeZB) [inline iphone/Classes/TiUIiOSApplicationShortcutsProxy.m]: You need to strip the first slash first or `/images/myIcon.png` won't work.
See https://jira.appcelerator.org/browse/TIMOB-19709?focusedCommentId=366605&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-366605

**Comment by FokkeZB** [comment]
@hansemannn added some comments to the code.

Also, the title of this PR should have all JIRA tickets in separate brackets or they won't be linked from the JIRA ticket.

**Comment by FokkeZB** [comment]
Updated https://github.com/appcelerator-developer-relations/appc-sample-3dtouch

File: iphone/iphone/Titanium.plist (hunk 1 above)
Comment (hansemannn) [inline iphone/iphone/Titanium.plist]: Will be ignored on non-supported devices and improves KitchenSink testing, so its ok to leave it there.

File: iphone/Classes/TiUIiOSApplicationShortcutsProxy.m (hunk 2 above)
Comment (hansemannn) [inline iphone/Classes/TiUIiOSApplicationShortcutsProxy.m]: Not yet possible, as it returns something different.

File: apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml (hunk 3 above)
Comment (hansemannn) [inline apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml]: For debug, will remove that lines.

File: apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml (hunk 4 above)
Comment (hansemannn) [inline apidoc/Titanium/UI/iOS/ApplicationShortcuts.yml]: Thanks!

File: iphone/iphone/Titanium.plist (hunk 1 above)
Comment (FokkeZB) [inline iphone/iphone/Titanium.plist]: But on devices that do support it, won't it make this shortcut appear without the developer expecting it to?

**Comment by pec1985** [comment]
PR Reviewed. **APPROVED!**

**Pull request merged**